### DISCUSSION TITLE
router.push() progress bar

### DISCUSSION THREAD

**Discussion by rossanodr** [vercel/next.js#46478] [Ideas]
### Describe the feature you'd like to request

A progress bar which shows the progress while loading a new page.
It's a little strange to use useRouter and have no indication that the new page is going to load.

### Describe the solution you'd like

Something like https://www.npmjs.com/package/nextjs-progressbar
It was great.

### Describe alternatives you've considered

Currently in next Js 13  unfortunately I've found no alternatives

**Comment by rossanodr** [comment]
> You can play an animation on the `routeChangeStart'` event and end it on the `routeChangeEnd` event
> 
> ```js
>   const router = useRouter()
>   useEffect(() => {
>     router.events.on('routeChangeStart', routeChangeStart);
>     router.events.on('routeChangeComplete', routeChangeEnd);
>     router.events.on('routeChangeError', routeChangeError);
>     return () => {
>       router.events.off('routeChangeStart', routeChangeStart);
>       router.events.off('routeChangeComplete', routeChangeEnd);
>       router.events.off('routeChangeError', routeChangeError);
>     };
>   }, []);
> ```
> 
> This is how this library is doing actually: https://github.com/apal21/nextjs-progressbar/blob/master/src/index.tsx
> 
> Do you think something else is needed or you can just use the router events?

It does not work with next13.
the  useRouter from "next/navigation" does not have events :/ 
and the next/router can't be used with the app folder

**Comment by SamRoehrich** [comment]
I am curious if there is a way to use Suspense to achieve this. Not sure what the requirements are for your use case but I could see if you had 3 requests for example, updating the progress of a loader component as each request comes back.

**Comment by JavierLopezSCG** [comment]
Bumping the issue.

The new useRouter from Next 13 'next/navigation' (AppRouterInstance) doesn't expose any properties to detect the events on routes. Any possible solution?

**Comment by guitheengineer** [comment]
I think they'll eventually add these events again, just a matter of time. 

One PR is open to listen to routeChangeStart:
https://github.com/vercel/next.js/pull/46391

Edit: PR got closed
![image](https://user-images.githubusercontent.com/49620737/221606455-408f5457-b683-4a1e-a969-66f939d5f8c2.png)

**Comment by kachkaev** [comment]
@lforst @timneutkens now that https://github.com/vercel/next.js/pull/46391 is closed, do you know any workarounds we can use? I am trying `/app` in one project and the lack of router events prevents me from using [`nprogress`](https://www.npmjs.com/package/nprogress). When I navigate between two pages with react server components locally, navigation is instant so UX feels fine. That's not the case in production on a slow network though, partly because there is no link preloading. Lack of UI feedback between a click and an update feels somewhat confusing.

Being able to show something like `nprogress` during navigation would be quite important. It might be possible to leverage native navigation events (https://github.com/apal21/nextjs-progressbar/issues/86#issuecomment-1305738335), but it is [not yet supported by FF and Safari](https://developer.mozilla.org/en-US/docs/Web/API/Navigation/navigate_event#browser_compatibility).

**Reply by lforst** [reply]
We would love to have a workaround for the Sentry SDK too. I was surprised/disappointed that this PR received pushback tbh. The use-case is IMO very valid.

**Reply by kachkaev** [reply]
I found workarounds for `nprogress` here:
- https://github.com/vercel/next.js/discussions/41745#discussioncomment-4208449
- https://github.com/vercel/next.js/discussions/41745#discussioncomment-4202641

And here is a recommendation on how to catch route change events:
- https://github.com/vercel/next.js/discussions/41745#discussioncomment-3987025

**Reply by lforst** [reply]
Interesting. One approach, for now, could be to monkey-patch the Link component and `router.push`.

**Comment by GoudekettingRM** [comment]
Is there any update on this?

**Reply by underscoDee** [reply]
Unfortunately no

**Comment by TudorEsan** [comment]
wow a year has passed but no solution yet

**Reply by kachkaev** [reply]
You can find a workaround here: https://github.com/TheSGJ/nextjs-toploader/issues/56

If I understand correctly, Next.js App router is not meant to work with a “classic” progress bar at the top of the whole page. Instead, the suggested approach is to use Suspense boundaries in different parts of the layout/page tree. When you click on a link, the skeleton of another page opens instantly but most of it is suspend (unless preloaded). Then these suspense ‘islands’ progressively load bit by bit, producing a complete UI in the end.

Showing a single progress bar above the entire page does not add much value because there is enough Suspense feedback in the UI already. You won't find this progress bar at the top of facebook web app, for example.

**Reply by korospace** [reply]
I think I found the solution here

```js
// import { useRouter } from "next/navigation";
import { useRouter } from "next-nprogress-bar";

const router = useRouter();

router.push("/home");
```

**Reply by anhphamduy** [reply]
tysm, works for me

**Reply by dan-pugsley** [reply]
> I think I found the solution here
> 
> ```js
> // import { useRouter } from "next/navigation";
> import { useRouter } from "next-nprogress-bar";
> 
> const router = useRouter();
> 
> router.push("/home");
> ```

Amazing :) that works!

**Reply by rezha4** [reply]
if anyone is using next-intl or any other custom useRouter, this approach works:

```typescript
import { useRouter as i18nRouter } from "@/i18n/routing";
import { useRouter } from "next-nprogress-bar";

const router = useRouter(i18nRouter);
```

it's mentioned in the npm [docs](https://www.npmjs.com/package/next-nprogress-bar#user-content-use-1)

**Comment by thehadikarimi** [comment]
Hey everyone,

I actually just released a package called `nextjs-progress` that does exactly this, and it’s built to work seamlessly with both Next.js **App Router** and **Pages Router** directories!

It provides a customizable progress bar that shows during navigation, giving users that visual feedback they expect. It also handles back/forward and same-page navigations.

Also you can create you own progress indicator if you do not want to use default progress bar.

You can find it here: [nextjs-progress](https://github.com/thehadikarimi/nextjs-progress)

Hope this is a helpful alternative for anyone looking for this feature!
